src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -153,6 +154,51 @@ public static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropRe
         return target;
     }
 
+    private static MethodType computeCollectArgumentsType(MethodType target, int insertIndex, MethodType collector) {
+        if (collector.returnType() != void.class){
+            // for non-void collector, the returned value is inserted into the argument list passed to the target
+            target = target.dropParameterTypes(insertIndex, insertIndex + 1);
+        }
+        return target.insertParameterTypes(insertIndex, collector.parameterArray()); // insert targetType parameters instead
+    }
+
+    // this just computes the adjusted type
+    private static MethodType computeUpcallIMRType(MethodType targetType, boolean dropReturn) {
+        if (targetType.returnType() != MemorySegment.class)
+            throw new IllegalArgumentException("Must return MemorySegment for IMR");
+
+        targetType = computeCollectArgumentsType(MH_BUFFER_COPY.type(), 1, targetType);
+
+        if (dropReturn) { // no handling for return value, need to drop it
+            targetType = targetType.changeReturnType(void.class);
+        } else {
+            // adjust return type so it matches the inferred type of the effective
+            // function descriptor
+            targetType = targetType.changeReturnType(MemorySegment.class);
+        }
+
+        return targetType;
+    }
+
+    public static UpcallStubFactory arrangeUpcallHelper(MethodType targetType, boolean isInMemoryReturn, boolean dropReturn,
+                                                        ABIDescriptor abi, CallingSequence callingSequence) {
+        if (isInMemoryReturn) {
+            targetType = computeUpcallIMRType(targetType, dropReturn);
+        }
+
+        UpcallStubFactory factory = UpcallLinker.makeFactory(targetType, abi, callingSequence);
+
+        if (isInMemoryReturn) {
```

Comment:
Question: does this logic belong here? E.g. we have a method that is used to create an upcall factory, namely `UpcallLinker.makeFactory`, but it turns out that if we have `isInMemoryReturn` some other factory should be used instead. Shouldn't we tweak the code so that `makeFactory` always does the right thing?

Reply:
It's a matter of organization. I think we should keep `UpcallLinker` ABI agnostic, and in memory returns are figments of an ABI. (note e.g. that the `CallingSequence` we pass to `makeFactory` has no notion of this in memory return. Not to be confused with the return buffer)

I see the additional step for in memory return not as using another factory, but as providing some pre-processing of the MethodHandle, before it's passed to `makeFactory`. The pre-processing is ABI-specific, so should live outside of UpcallLinker, IMO.